File: src/transports/http-transport.ts
/**
 * HTTP Transport Implementation
 * Sends events to TraceFlow REST API with retry logic and circuit breaker
 */

import {
  TraceEvent,
  TraceEventType,
  TraceTransport,
  TraceStatus,
  StepStatus,
  HTTPTracePayload,
  HTTPStepPayload,
  HTTPLogPayload,
} from '../types';

export interface HTTPTransportConfig {
  endpoint: string;
  apiKey?: string;
  username?: string;
  password?: string;
  timeout?: number;
  maxRetries?: number;
  retryDelay?: number;
  enableCircuitBreaker?: boolean;
  silentErrors?: boolean;
}

interface QueuedRequest {
  url: string;
  method: string;
  body: any;
  retries: number;
}

/**
 * HTTP Transport with exponential backoff, circuit breaker, and batching
 */
export class HTTPTransport implements TraceTransport {
  private config: Required<HTTPTransportConfig>;
  private queue: QueuedRequest[] = [];
  private circuitOpen: boolean = false;
  private circuitOpenUntil: number = 0;
  private failureCount: number = 0;
  private readonly CIRCUIT_THRESHOLD = 5;
  private readonly CIRCUIT_TIMEOUT = 60000; // 1 minute

  constructor(config: HTTPTransportConfig) {
    this.config = {
      endpoint: config.endpoint,
      apiKey: config.apiKey || '',
      username: config.username || '',
      password: config.password || '',
      timeout: config.timeout || 5000,
      maxRetries: config.maxRetries || 3,
      retryDelay: config.retryDelay || 1000,
      enableCircuitBreaker: config.enableCircuitBreaker ?? true,
      silentErrors: config.silentErrors ?? true,
    };
  }

  /**
   * Send event to HTTP API
   */
  async send(event: TraceEvent): Promise<void> {
    try {
      // Check circuit breaker
      if (this.isCircuitOpen()) {
        if (this.config.silentErrors) {
          console.warn(`[TraceFlow HTTP] Circuit open, queueing event: ${event.event_type}`);
          return;
        }
        throw new Error('Circuit breaker is open');
      }

      // Convert event to HTTP payload and send
      await this.sendEventToAPI(event);
      
      // Reset failure count on success
      this.failureCount = 0;
    } catch (error) {
      this.handleError(error, event);
    }
  }

  /**
   * Flush any pending events
   */
  async flush(): Promise<void> {
    if (this.queue.length === 0) {
      return;
    }

    console.log(`[TraceFlow HTTP] Flushing ${this.queue.length} queued requests...`);
    
    const requests = [...this.queue];
    this.queue = [];

    for (const req of requests) {
      try {
        await this.executeRequest(req.url, req.method, req.body);
      } catch (error) {
        if (!this.config.silentErrors) {
          console.error('[TraceFlow HTTP] Failed to flush request:', error);
        }
      }
    }
  }

  /**
   * Shutdown transport gracefully
   */
  async shutdown(): Promise<void> {
    await this.flush();
  }

  /**
   * Send event to appropriate API endpoint
   */
  private async sendEventToAPI(event: TraceEvent): Promise<void> {
    switch (event.event_type) {
      case TraceEventType.TRACE_STARTED:
        await this.createTrace(event);
        break;
      
      case TraceEventType.TRACE_FINISHED:
      case TraceEventType.TRACE_FAILED:
      case TraceEventType.TRACE_CANCELLED:
        await this.updateTrace(event);
        break;
      
      case TraceEventType.STEP_STARTED:
        await this.createStep(event);
        break;
      
      case TraceEventType.STEP_FINISHED:
      case TraceEventType.STEP_FAILED:
        await this.updateStep(event);
        break;
      
      case TraceEventType.LOG_EMITTED:
        await this.createLog(event);
        break;
      
      default:
        console.warn(`[TraceFlow HTTP] Unknown event type: ${event.event_type}`);
    }
  }

  /**
   * Create trace via POST /api/v1/traces
   */
  private async createTrace(event: TraceEvent): Promise<void> {
    const payload: HTTPTracePayload = {
      trace_id: event.trace_id,
      trace_type: event.payload.trace_type,
      status: TraceStatus.PENDING,
      source: event.source,
      created_at: event.timestamp,
      updated_at: event.timestamp,
      title: event.payload.title,
      description: event.payload.description,
      owner: event.payload.owner,
      tags: event.payload.tags,
      metadata: event.payload.metadata,
      params: event.payload.params,
      last_activity_at: event.timestamp,
      idempotency_key: event.payload.idempotency_key || event.event_id,
      trace_timeout_ms: event.payload.trace_timeout_ms,
      step_timeout_ms: event.payload.step_timeout_ms,
    };

    await this.executeRequestWithRetry(
      `${this.config.endpoint}/api/v1/traces`,
      'POST',
      payload
    );
  }

  /**
   * Update trace via PATCH /api/v1/traces/{id}
   */
  private async updateTrace(event: TraceEvent): Promise<void> {
    let status: TraceStatus;
    
    switch (event.event_type) {
      case TraceEventType.TRACE_FINISHED:
        status = TraceStatus.SUCCESS;
        break;
      case TraceEventType.TRACE_FAILED:
        status = TraceStatus.FAILED;
        break;
      case TraceEventType.TRACE_CANCELLED:
        status = TraceStatus.CANCELLED;
        break;
      default:
        status = TraceStatus.RUNNING;
    }

    const payload: Partial<HTTPTracePayload> = {
      status,
      updated_at: event.timestamp,
      finished_at: event.timestamp,
      last_activity_at: event.timestamp,
      result: event.payload.result,
      error: event.payload.error,
      metadata: event.payload.metadata,
    };

    await this.executeRequestWithRetry(
      `${this.config.endpoint}/api/v1/traces/${event.trace_id}`,
      'PATCH',
      payload
    );
  }

  /**
   * Create step via POST /api/v1/steps
   */
  private async createStep(event: TraceEvent): Promise<void> {
    const payload: HTTPStepPayload = {
      trace_id: event.trace_id,
      step_id: event.step_id!,
      step_type: event.payload.step_type,
      name: event.payload.name,
      status: StepStatus.STARTED,
      started_at: event.timestamp,
      updated_at: event.timestamp,
      input: event.payload.input,
      metadata: event.payload.metadata,
    };

    await this.executeRequestWithRetry(
      `${this.config.endpoint}/api/v1/steps`,
      'POST',
      payload
    );
  }

  /**
   * Update step via PATCH /api/v1/steps/{traceId}/{stepId}
   */
  private async updateStep(event: TraceEvent): Promise<void> {
    const status = event.event_type === TraceEventType.STEP_FINISHED
      ? StepStatus.COMPLETED
      : StepStatus.FAILED;

    const payload: Partial<HTTPStepPayload> = {
      status,
      updated_at: event.timestamp,
      finished_at: event.timestamp,
      output: event.payload.output,
      error: event.payload.error,
      metadata: event.payload.metadata,
    };

    await this.executeRequestWithRetry(
      `${this.config.endpoint}/api/v1/steps/${event.trace_id}/${event.step_id}`,
      'PATCH',
      payload
    );
  }

  /**
   * Create log via POST /api/v1/logs
   */
  private async createLog(event: TraceEvent): Promise<void> {
    const payload: HTTPLogPayload = {
      trace_id: event.trace_id,
      log_time: event.timestamp,
      log_id: event.event_id,
      step_number: event.payload.step_number,
      level: event.payload.level,
      event_type: event.payload.event_type,
      message: event.payload.message,
      details: event.payload.details,
      source: event.source,
    };

    await this.executeRequestWithRetry(
      `${this.config.endpoint}/api/v1/logs`,
      'POST',
      payload
    );
  }

  /**
   * Execute HTTP request with retry logic
   */
  private async executeRequestWithRetry(
    url: string,
    method: string,
    body: any,
    retries: number = 0
  ): Promise<void> {
    try {
      await this.executeRequest(url, method, body);
    } catch (error: any) {
      // Retry on network errors or 5xx status codes
      const shouldRetry = 
        retries < this.config.maxRetries &&
        (this.isNetworkError(error) || this.is5xxError(error));

      if (shouldRetry) {
        const delay = this.calculateBackoff(retries);
        console.warn(`[TraceFlow HTTP] Retry ${retries + 1}/${this.config.maxRetries} after ${delay}ms`);
        
        await this.sleep(delay);
        return this.executeRequestWithRetry(url, method, body, retries + 1);
      }

      // Max retries exceeded, handle failure
      this.recordFailure();
      
      if (!this.config.silentErrors) {
        throw error;
      }
      
      console.error('[TraceFlow HTTP] Request failed after retries:', error.message);
    }
  }

  /**
   * Execute single HTTP request
   */
  private async executeRequest(url: string, method: string, body: any): Promise<void> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };

    // Add authentication
    if (this.config.apiKey) {
      headers['X-API-Key'] = this.config.apiKey;
    } else if (this.config.username && this.config.password) {
      const auth = Buffer.from(`${this.config.username}:${this.config.password}`).toString('base64');
      headers['Authorization'] = `Basic ${auth}`;
    }

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);

    try {
      const response = await fetch(url, {
        method,
        headers,
        body: JSON.stringify(body),
        signal: controller.signal,
      });

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }
    } finally {
      clearTimeout(timeoutId);
    }
  }

  /**
   * Calculate exponential backoff with jitter
   */
  private calculateBackoff(retries: number): number {
    const exponentialDelay = this.config.retryDelay * Math.pow(2, retries);
    const jitter = Math.random() * 1000;
    return Math.min(exponentialDelay + jitter, 30000); // Max 30 seconds
  }

  /**
   * Check if error is a network error
   */
  private isNetworkError(error: any): boolean {
    return (
      error.name === 'AbortError' ||
      error.name === 'NetworkError' ||
      error.code === 'ECONNREFUSED' ||
      error.code === 'ETIMEDOUT' ||
      error.code === 'ENOTFOUND'
    );
  }

  /**
   * Check if error is a 5xx server error
   */
  private is5xxError(error: any): boolean {
    const message = error.message || '';
    return /HTTP 5\d{2}/.test(message);
  }

  /**
   * Record failure for circuit breaker
   */
  private recordFailure(): void {
    if (!this.config.enableCircuitBreaker) {
      return;
    }

    this.failureCount++;
    
    if (this.failureCount >= this.CIRCUIT_THRESHOLD) {
      this.circuitOpen = true;
      this.circuitOpenUntil = Date.now() + this.CIRCUIT_TIMEOUT;
      console.warn(`[TraceFlow HTTP] Circuit breaker opened for ${this.CIRCUIT_TIMEOUT}ms`);
    }
  }

  /**
   * Check if circuit breaker is open
   */
  private isCircuitOpen(): boolean {
    if (!this.config.enableCircuitBreaker) {
      return false;
    }

    if (this.circuitOpen && Date.now() > this.circuitOpenUntil) {
      console.log('[TraceFlow HTTP] Circuit breaker closed, resuming requests');
      this.circuitOpen = false;
      this.failureCount = 0;
    }

    return this.circuitOpen;
  }

  /**
   * Handle errors gracefully
   */
  private handleError(error: any, event: TraceEvent): void {
    if (this.config.silentErrors) {
      console.error('[TraceFlow HTTP] Error sending event (silenced):', error.message);
    } else {
      throw error;
    }
  }

  /**
   * Sleep utility
   */
  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

File: go/http_transport.go
package traceflow

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"math"
	"math/rand"
	"net/http"
	"regexp"
	"sync"
	"time"
)

// httpTransportConfig configures the HTTP transport.
type httpTransportConfig struct {
	Endpoint                string
	APIKey                  string
	Username                string
	Password                string
	Timeout                 time.Duration
	MaxRetries              int
	RetryDelay              time.Duration
	EnableCircuitBreaker    bool
	CircuitBreakerThreshold int
	CircuitBreakerTimeout   time.Duration
	SilentErrors            bool
}

var serverErrorRegex = regexp.MustCompile(`HTTP 5\d{2}`)

// httpTransport delivers events to the TraceFlow REST API with exponential
// backoff retries and a circuit breaker. Events arriving while the circuit is
// open are buffered and replayed when it closes.
type httpTransport struct {
	cfg    httpTransportConfig
	client *http.Client
	logger Logger

	mu             sync.Mutex
	circuitOpen    bool
	circuitOpenAt  time.Time
	failureCount   int
	pendingEvents  []TraceEvent
	circuitTimeout time.Duration
	threshold      int
}

func newHTTPTransport(cfg httpTransportConfig, logger Logger) *httpTransport {
	if cfg.Timeout == 0 {
		cfg.Timeout = 5 * time.Second
	}
	// MaxRetries is used verbatim: 0 means no retries. The production default
	// (3) is applied by Client.New so it can be overridden, while callers that
	// construct the transport directly retain full control.
	if cfg.RetryDelay == 0 {
		cfg.RetryDelay = time.Second
	}
	if cfg.CircuitBreakerThreshold == 0 {
		cfg.CircuitBreakerThreshold = 5
	}
	if cfg.CircuitBreakerTimeout == 0 {
		cfg.CircuitBreakerTimeout = 60 * time.Second
	}
	if logger == nil {
		logger = noopLogger{}
	}
	return &httpTransport{
		cfg:            cfg,
		client:         &http.Client{Timeout: cfg.Timeout},
		logger:         logger,
		circuitTimeout: cfg.CircuitBreakerTimeout,
		threshold:      cfg.CircuitBreakerThreshold,
	}
}

func (t *httpTransport) Send(ctx context.Context, event TraceEvent) error {
	if t.isCircuitOpen() {
		t.mu.Lock()
		t.pendingEvents = append(t.pendingEvents, event)
		pending := len(t.pendingEvents)
		t.mu.Unlock()
		t.logger.Warn("Circuit open, queued event: %s (%d pending)", event.EventType, pending)
		if !t.cfg.SilentErrors {
			return fmt.Errorf("circuit breaker is open")
		}
		return nil
	}

	if err := t.sendEventToAPI(ctx, event); err != nil {
		if t.cfg.SilentErrors {
			t.logger.Error("Error sending event (silenced): %v", err)
			return nil
		}
		return err
	}
	return nil
}

func (t *httpTransport) Flush(ctx context.Context) error {
	t.mu.Lock()
	pending := t.pendingEvents
	t.pendingEvents = nil
	t.mu.Unlock()

	if len(pending) > 0 {
		t.logger.Info("Flushing %d circuit-breaker-queued events...", len(pending))
		for _, event := range pending {
			if err := t.sendEventToAPI(ctx, event); err != nil && !t.cfg.SilentErrors {
				t.logger.Error("Failed to flush pending event: %v", err)
			}
		}
	}
	return nil
}

func (t *httpTransport) Shutdown(ctx context.Context) error {
	return t.Flush(ctx)
}

func (t *httpTransport) HealthCheck(ctx context.Context) HealthCheckResult {
	start := time.Now()
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, t.cfg.Endpoint+"/api/v1/health", nil)
	if err != nil {
		return HealthCheckResult{OK: false, LatencyMs: time.Since(start).Milliseconds(), Error: err.Error()}
	}
	t.applyAuth(req)
	resp, err := t.client.Do(req)
	latency := time.Since(start).Milliseconds()
	if err != nil {
		return HealthCheckResult{OK: false, LatencyMs: latency, Error: err.Error()}
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return HealthCheckResult{OK: false, LatencyMs: latency, Error: fmt.Sprintf("HTTP %d: %s", resp.StatusCode, resp.Status)}
	}
	return HealthCheckResult{OK: true, LatencyMs: latency}
}

func (t *httpTransport) sendEventToAPI(ctx context.Context, event TraceEvent) error {
	switch event.EventType {
	case EventTraceStarted:
		return t.createTrace(ctx, event)
	case EventTraceFinished, EventTraceFailed, EventTraceCancelled:
		return t.updateTrace(ctx, event)
	case EventStepStarted:
		return t.createStep(ctx, event)
	case EventStepFinished, EventStepFailed:
		return t.updateStep(ctx, event)
	case EventLogEmitted:
		return t.createLog(ctx, event)
	default:
		t.logger.Warn("Unknown event type: %s", event.EventType)
		return nil
	}
}

func (t *httpTransport) createTrace(ctx context.Context, event TraceEvent) error {
	idempotency, _ := event.Payload["idempotency_key"].(string)
	if idempotency == "" {
		idempotency = event.EventID
	}
	payload := httpTracePayload{
		TraceID:        event.TraceID,
		TraceType:      asString(event.Payload["trace_type"]),
		Status:         TraceStatusPending,
		Source:         event.Source,
		CreatedAt:      event.Timestamp,
		UpdatedAt:      event.Timestamp,
		Title:          asString(event.Payload["title"]),
		Description:    asString(event.Payload["description"]),
		Owner:          asString(event.Payload["owner"]),
		Tags:           asStringSlice(event.Payload["tags"]),
		Metadata:       asMap(event.Payload["metadata"]),
		Params:         event.Payload["params"],
		LastActivityAt: event.Timestamp,
		IdempotencyKey: idempotency,
		TraceTimeoutMs: asInt(event.Payload["trace_timeout_ms"]),
		StepTimeoutMs:  asInt(event.Payload["step_timeout_ms"]),
	}
	return t.executeWithRetry(ctx, http.MethodPost, t.cfg.Endpoint+"/api/v1/traces", payload)
}

func (t *httpTransport) updateTrace(ctx context.Context, event TraceEvent) error {
	var status TraceStatus
	switch event.EventType {
	case EventTraceFinished:
		status = TraceStatusSuccess
	case EventTraceFailed:
		status = TraceStatusFailed
	case EventTraceCancelled:
		status = TraceStatusCancelled
	default:
		status = TraceStatusRunning
	}
	payload := httpTracePayload{
		Status:         status,
		UpdatedAt:      event.Timestamp,
		FinishedAt:     event.Timestamp,
		LastActivityAt: event.Timestamp,
		Result:         event.Payload["result"],
		Error:          asString(event.Payload["error"]),
		Metadata:       asMap(event.Payload["metadata"]),
	}
	return t.executeWithRetry(ctx, http.MethodPatch, t.cfg.Endpoint+"/api/v1/traces/"+event.TraceID, payload)
}

func (t *httpTransport) createStep(ctx context.Context, event TraceEvent) error {
	payload := httpStepPayload{
		TraceID:   event.TraceID,
		StepID:    event.StepID,
		StepType:  asString(event.Payload["step_type"]),
		Name:      asString(event.Payload["name"]),
		Status:    StepStatusStarted,
		StartedAt: event.Timestamp,
		UpdatedAt: event.Timestamp,
		Input:     event.Payload["input"],
		Metadata:  asMap(event.Payload["metadata"]),
	}
	return t.executeWithRetry(ctx, http.MethodPost, t.cfg.Endpoint+"/api/v1/steps", payload)
}

func (t *httpTransport) updateStep(ctx context.Context, event TraceEvent) error {
	status := StepStatusFailed
	if event.EventType == EventStepFinished {
		status = StepStatusCompleted
	}
	payload := httpStepPayload{
		Status:     status,
		UpdatedAt:  event.Timestamp,
		FinishedAt: event.Timestamp,
		Output:     event.Payload["output"],
		Error:      asString(event.Payload["error"]),
		Metadata:   asMap(event.Payload["metadata"]),
	}
	url := fmt.Sprintf("%s/api/v1/steps/%s/%s", t.cfg.Endpoint, event.TraceID, event.StepID)
	return t.executeWithRetry(ctx, http.MethodPatch, url, payload)
}

func (t *httpTransport) createLog(ctx context.Context, event TraceEvent) error {
	payload := httpLogPayload{
		TraceID:   event.TraceID,
		LogTime:   event.Timestamp,
		LogID:     event.EventID,
		Level:     asString(event.Payload["level"]),
		EventType: asString(event.Payload["event_type"]),
		Message:   asString(event.Payload["message"]),
		Details:   event.Payload["details"],
		Source:    event.Source,
	}
	return t.executeWithRetry(ctx, http.MethodPost, t.cfg.Endpoint+"/api/v1/logs", payload)
}

func (t *httpTransport) executeWithRetry(ctx context.Context, method, url string, body any) error {
	var lastErr error
	for attempt := 0; ; attempt++ {
		err := t.execute(ctx, method, url, body)
		if err == nil {
			t.mu.Lock()
			t.failureCount = 0
			t.mu.Unlock()
			return nil
		}
		lastErr = err

		if attempt < t.cfg.MaxRetries && (isNetworkError(err) || isServerError(err)) {
			delay := t.backoff(attempt)
			t.logger.Warn("Retry %d/%d after %s", attempt+1, t.cfg.MaxRetries, delay)
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(delay):
			}
			continue
		}

		t.recordFailure()
		if !t.cfg.SilentErrors {
			return lastErr
		}
		t.logger.Error("Request failed after retries: %v", lastErr)
		return nil
	}
}

func (t *httpTransport) execute(ctx context.Context, method, url string, body any) error {
	data, err := json.Marshal(sanitize(body))
	if err != nil {
		return fmt.Errorf("marshal payload: %w", err)
	}
	req, err := http.NewRequestWithContext(ctx, method, url, bytes.NewReader(data))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	t.applyAuth(req)

	resp, err := t.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("HTTP %d: %s", resp.StatusCode, resp.Status)
	}
	return nil
}

func (t *httpTransport) applyAuth(req *http.Request) {
	switch {
	case t.cfg.APIKey != "":
		req.Header.Set("X-API-Key", t.cfg.APIKey)
	case t.cfg.Username != "" && t.cfg.Password != "":
		auth := base64.StdEncoding.EncodeToString([]byte(t.cfg.Username + ":" + t.cfg.Password))
		req.Header.Set("Authorization", "Basic "+auth)
	}
}

func (t *httpTransport) backoff(attempt int) time.Duration {
	exponential := float64(t.cfg.RetryDelay) * math.Pow(2, float64(attempt))
	jitter := rand.Float64() * float64(time.Second)
	d := time.Duration(exponential + jitter)
	if max := 30 * time.Second; d > max {
		return max
	}
	return d
}

func (t *httpTransport) recordFailure() {
	if !t.cfg.EnableCircuitBreaker {
		return
	}
	t.mu.Lock()
	defer t.mu.Unlock()
	t.failureCount++
	if t.failureCount >= t.threshold {
		t.circuitOpen = true
		t.circuitOpenAt = time.Now().Add(t.circuitTimeout)
		t.logger.Warn("Circuit breaker opened for %s", t.circuitTimeout)
	}
}

func (t *httpTransport) isCircuitOpen() bool {
	if !t.cfg.EnableCircuitBreaker {
		return false
	}
	t.mu.Lock()
	defer t.mu.Unlock()
	if t.circuitOpen && time.Now().After(t.circuitOpenAt) {
		t.logger.Info("Circuit breaker closed, resuming requests")
		t.circuitOpen = false
		t.failureCount = 0
		t.drainLocked()
	}
	return t.circuitOpen
}

// drainLocked replays buffered events asynchronously. Caller holds t.mu.
func (t *httpTransport) drainLocked() {
	if len(t.pendingEvents) == 0 {
		return
	}
	events := t.pendingEvents
	t.pendingEvents = nil
	t.logger.Info("Draining %d pending events after circuit close", len(events))
	for _, event := range events {
		go func(e TraceEvent) {
			if err := t.sendEventToAPI(context.Background(), e); err != nil {
				t.logger.Error("Failed to drain pending event: %v", err)
			}
		}(event)
	}
}

func isNetworkError(err error) bool {
	// http.Client surfaces timeouts, connection refused, DNS failures, and
	// context cancellation as *url.Error wrapping the cause; treat anything
	// that is not an explicit HTTP status error as retryable transport noise.
	return err != nil && !serverErrorRegex.MatchString(err.Error()) && !is4xx(err)
}

func isServerError(err error) bool {
	return err != nil && serverErrorRegex.MatchString(err.Error())
}

func is4xx(err error) bool {
	return err != nil && regexp.MustCompile(`HTTP 4\d{2}`).MatchString(err.Error())
}
